src/jacobi/_jacobi.py

```diff
@@ -1 +1,261 @@
+            fd.append(np.reshape(fdi, -1) if i == 1 else fdi[todo])
+            if diagnostic is not None:
+                diagnostic["call"][todo, k] += 2
+                diagnostic["iteration"][k] += 1
+
+            # polynomial fit with one extra degree of freedom;
+            # use latest maxgrad + 1 data points
+            grad = min(i - 1, maxgrad)
+            start = i - (grad + 1)
+            stop = i + 1
+            q, c = np.polyfit(
+                h[start:stop] ** 2, fd[start:], grad, rcond=None, cov=True
+            )
+            ri = q[-1]
+            # pulls have roughly unit variance, however,
+            # the pull distribution is not gaussian and looks
+            # more like student's t
+            rei = c[-1, -1] ** 0.5
+
+            # update estimates that have smaller estimated error
+            sub_todo = rei < re[todo]
+            todo1 = todo.copy()
+            todo[todo1] = sub_todo
+            r[todo] = ri[sub_todo]
+            re[todo] = rei[sub_todo]
+
+            # do not improve estimates further which meet the tolerance
+            if rtol > 0:
+                sub_todo &= rei > rtol * np.abs(ri)
+                todo[todo1] = sub_todo
+
+            if diagnostic is not None:
+                re2 = re.copy()
+                re2[todo1] = rei
+                diagnostic["residual"][k].append(re2)
+
+            if np.sum(todo) == 0:
+                break
+
+            # shrink previous vectors of estimates
+            fd = [fdi[sub_todo] for fdi in fd]
+
+        if jac.ndim == 0:
+            jac[...] = r
+            err[...] = re
+        elif jac.ndim == 1:
+            jac[kx] = r
+            err[kx] = re
+        else:
+            jac[(...,) + kx] = r
+            err[(...,) + kx] = re
+
+        it.iternext()
+
+    return jac, err
+
+
+def _steps(p, step, maxiter):
+    h0, factor = step
+    h = p * h0
+    if not h != 0:  # also works if p is NaN
+        h = h0
+    return h * factor ** np.arange(maxiter)
+
+
+def _derive(mode, f0, f, x, i, h, args):
+    x1 = x.copy()
+    x2 = x.copy()
+    if mode == 0:
+        x1[i] += h
+        x2[i] -= h
+        return (f(x1, *args) - f(x2, *args)) * (0.5 / h)
+    h = h * mode
+    x1[i] += h
+    x2[i] += 2 * h
+    f1 = f(x1, *args)
+    f2 = f(x2, *args)
+    return (-3 * f0 + 4 * f1 - f2) * (0.5 / h)
+
+
+def _first(method, f0, f, x, i, h, args):
+    norm = 0.5 / h
+    f1 = None
+    f2 = None
+    if method is None or method == 0:
+        x1 = x.copy()
+        x2 = x.copy()
+        x1[i] -= h
+        x2[i] += h
+        f1 = f(x1, *args)
+        f2 = f(x2, *args)
+        if method is None:
+            if np.any(np.isnan(f1)):  # forward method
+                method = 1
+            elif np.any(np.isnan(f2)):  # backward method
+                method = -1
+            else:
+                method = 0
+    if method == 0:
+        return method, None, (f1 - f2) * norm
```

Comment:
```suggestion
        return method, None, (f2 - f1) * norm
```
I think this produces the negative of the derivative estimate as-is.

Reply:
No this should be correct. I always compute f1 - f2 and x1 - x2 consistently everywhere, so h should also be negative for x1 < x2. I hope that one of the unit tests breaks if you made that change.

Reply:
I'd suggest taking a look; there are only a few lines to follow starting at 121 then jumping into this function. I noticed the numerical effect before finding the cause here. I saw that the `_first` derivative estimate changes sign depending on whether it's doing central or one-sided differences.

I haven't run the tests because I didn't install this properly (just copy-pasted), but I wouldn't be surprised if they didn't catch it. It only affects the first estimate because every other iteration uses `_derive`.